debian/pwndck.1.md
---
date: February 2026
section: 1
title: PWNDCK
---

# NAME

pwndck - Check passwords against the HaveIBeenPwned leak database

# DESCRIPTION
 
usage: pwndck \[-q\] \[ \[\<password\> \...\] \| -i \<INPUT\> \| -e \| \--version \| -h \]

Report \# of password hits in HaveIBeenPwned

## positional arguments:

password

:   The password(s) to check

## options:

**-q**, **\--quiet**

:   suppress output

**-i**, **\--input** \<INPUT\>

:   file containing passwords, one per line (\'-\' for stdin)

**-e**, **\--estimatedb**

:   estimate the current size of the HaveIBeenPwnd password database

**\--version**

:   show program\'s version number and exit

**-h**, **\--help**

:   show this help message and exit

Evaluate one or more passwords against the HaveIBeenPwned password
database, and return the number of accounts for which they have been
reported as compromised.

The number of entries found in the database is returned. If multiple
passwords are being checked, the password name is also returned.

If the password is not specified on the command line, and there is no
std input, the user will be prompted.

The command returns with an error code if a password is found in the
database.

The process\_pw() function is available for use in Python. It can be
accessed by importing it from the pwndck module.

See https://haveibeenpwned.com/API/v3#PwnedPasswords


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.6.2] - 2026-06-03
- Fix main entry point

## [0.6.1] - 2026-06-03
- Cleanup/refactoring
- Documentation clarification

## [0.6] - 2026-02-22
- Added a HIBP password database size estimator
- Source refactoring

## [0.5.1] - 2026-02-22
- Documentation cleanup
- Additional testing

## [0.5] - 2026-02-12
- Rework API for module use
- Align columns in verbose mode
- Add devtest

## [0.4] - 2025-12-14
- Add testing
- Clean up copyrights

## [0.3] - 2025-12-04
- Support processing of multiple passwords
- Support a password input file
- General cleanup

## [0.2] - 2025-10-24
- Enable Brotli compression on HIBP requests
- Enable padding on HIBP responses for extra security

## [0.1] - 2025-10-20
- Original Release


pwndck/pwndck.py
#!/usr/bin/python3
#
# Copyright (c) 2025 David Steele <dsteele@gmail.com>
#
# SPDX-License-Identifier: GPL-2.0-or-later
# License-Filename: LICENSE
#

import argparse
import fileinput
import hashlib
import sys
import textwrap
import types
from typing import Iterable, List

import requests

from pwndck.flexi_formatter import FlexiHelpFormatter
from pwndck.version import __version__

apiurl = "https://api.pwnedpasswords.com/range/{}"


class PwndException(Exception):
    pass


def get_sha(data: str) -> str:
    hlib = hashlib.sha1()
    hlib.update(data.encode("utf-8"))
    hsh = hlib.hexdigest()

    return hsh.upper()


def get_hashes(key: str) -> str:
    """Return hash-adjacent results
    per https://haveibeenpwned.com/API/v3#PwnedPasswords"""

    url = apiurl.format(key)
    headers = {
        "User-Agent": f"PwndCk/{__version__}",
        "Add-Padding": "true",
    }
    r = requests.get(url, headers=headers)
    response = r.text

    if r.status_code != 200:
        raise PwndException(r.reason)

    return response


def process_pw(pw: str) -> int:
    """
    Returns the number of entries for a password in the Have I Been Pwned
    database.

    Parameters:
        pw (str): The password to check.

    Returns:
        int: The number of entries in the database.

    Raises:
        PwndException: For web query errors.
    """
    hsh = get_sha(pw)
    key = hsh[0:5]
    body = hsh[5:]

    for line in get_hashes(key).splitlines():
        if line.startswith(body):
            (body, count) = line.split(":")
            return int(count)

    return 0


def parse_args():
    parser = argparse.ArgumentParser(
        description="Report # of password hits in HaveIBeenPwned",
        epilog=textwrap.dedent(
            """
            Evaluate one or more passwords against the HaveIBeenPwned
            password database, and return the number of accounts for which
            they have been reported as compromised.

            The number of entries found in the database is returned. If
            multiple passwords are being checked, the password name is also
            returned.

            If the password is not specified on the command line, and there
            is no std input, the user will be prompted.

            The command returns with an error
            code if a password is found in the database.

            The process_pw() function is available for use in Python and can be
            accessed by importing it from its respective module.

            See https://haveibeenpwned.com/API/v3#PwnedPasswords
            """
        ),
        formatter_class=FlexiHelpFormatter,
    )

    parser.add_argument(
        "-q",
        "--quiet",
        help="suppress output",
        default=False,
        action="store_true",
    )

    group = parser.add_mutually_exclusive_group()

    group.add_argument(
        "-i",
        "--input",
        type=str,
        nargs="?",
        default=None,
        help="file containing passwords, one per line ('-' for stdin)",
    )

    group.add_argument(
        "passwords",
        help="The password(s) to check",
        nargs="*",
        default=None,
        type=str,
    )

    group.add_argument(
        "--version", action="version", version=f"%(prog)s {__version__}"
    )

    args = parser.parse_args()
    return args


def get_passwords(
    passwords_arg: List[str],
    input_file: str,
) -> Iterable[str]:
    if passwords_arg:
        return passwords_arg

    if input_file:
        return (
            f.strip()
            for f in fileinput.input(files=(input_file,), encoding="utf-8")
        )

    if sys.stdin.isatty():
        return [input("Enter password to check: ")]

    raise PwndException("No passwords")


def quiet_print(string, quiet) -> None:
    if not quiet:
        print(string)


def main() -> None:
    args = parse_args()

    try:
        passwords = get_passwords(args.passwords, args.input)

        fail = False
        verbose = isinstance(passwords, types.GeneratorType) or (
            isinstance(passwords, list) and len(passwords) > 1
        )
        for password in passwords:
            pwcount = process_pw(password)

            if verbose:
                quiet_print(f"{pwcount:<8}  {password:}", args.quiet)
            else:
                quiet_print(pwcount, args.quiet)

            if pwcount > 0:
                fail = True

    except FileNotFoundError:
        quiet_print("ERROR - Input file not found", args.quiet)
        sys.exit(-2)
    except PermissionError:
        quiet_print(
            "ERROR - Insufficient permissions for input file", args.quiet
        )
        sys.exit(-2)
    except KeyboardInterrupt:
        quiet_print("", args.quiet)
        sys.exit(-2)
    except PwndException as e:
        quiet_print(str(e), args.quiet)
        sys.exit(-2)

    if fail:
        sys.exit(-1)


if __name__ == "__main__":
    main()
